File: backend-java/src/main/java/mx/codemx/modules/evaluacion/EvaluacionRemotaStrategy.java
package mx.codemx.modules.evaluacion;

import java.time.Duration;
import java.util.List;
import mx.codemx.modules.retos.CasoPrueba;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.client.SimpleClientHttpRequestFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestClient;

/**
 * Estrategia concreta (Strategy): envía el código al Servicio Python externo (sandbox)
 * por HTTP/JSON con un timeout explícito. Si el servicio no responde, lanza excepción
 * para que el contexto (EvaluacionService) recurra a la estrategia de respaldo.
 */
@Component
public class EvaluacionRemotaStrategy implements EvaluacionStrategy {

    private final RestClient http;

    public EvaluacionRemotaStrategy(
            RestClient.Builder builder,
            @Value("${evaluador.url:http://localhost:8000}") String url,
            @Value("${evaluador.timeout-ms:5000}") int timeoutMs) {

        Duration timeout = Duration.ofMillis(timeoutMs);

        SimpleClientHttpRequestFactory factory = new SimpleClientHttpRequestFactory();
        factory.setConnectTimeout(timeout);
        factory.setReadTimeout(timeout);

        this.http = builder.baseUrl(url).requestFactory(factory).build();
    }

    @Override
    public ResultadoEvaluacion ejecutar(SolicitudEvaluacion solicitud, List<CasoPrueba> casos) {
        ResultadoEvaluacion resultado = http.post()
                .uri("/evaluar")
                .body(solicitud)
                .retrieve()
                .body(ResultadoEvaluacion.class);

        if (resultado == null) {
            throw new IllegalStateException("Respuesta vacía del Servicio Python");
        }
        return resultado;
    }
}

File: backend/Modules/Evaluacion/EvaluacionRemotaStrategy.cs
using System.Net.Http.Json;
using CodeMX.Api.Modules.Retos;

namespace CodeMX.Api.Modules.Evaluacion;

/// <summary>
/// Estrategia concreta (Strategy): envía el código al Servicio Python externo (sandbox)
/// por HTTP/JSON con un timeout explícito. Si el servicio no responde, lanza excepción
/// para que el contexto (EvaluacionService) recurra a la estrategia de respaldo.
/// </summary>
public class EvaluacionRemotaStrategy : IEvaluacionStrategy
{
    private readonly HttpClient _http;
    private readonly string _url;
    private readonly int _timeoutMs;

    public EvaluacionRemotaStrategy(HttpClient http, IConfiguration config)
    {
        _http = http;
        _url = config["Evaluador:Url"] ?? "http://localhost:8000";
        _timeoutMs = int.TryParse(config["Evaluador:TimeoutMs"], out var t) ? t : 5000;
    }

    public async Task<ResultadoEvaluacion> EjecutarAsync(SolicitudEvaluacion solicitud, List<CasoPrueba> casos)
    {
        using var cts = new CancellationTokenSource(_timeoutMs);
        var resp = await _http.PostAsJsonAsync($"{_url}/evaluar", solicitud, cts.Token);
        resp.EnsureSuccessStatusCode();
        return await resp.Content.ReadFromJsonAsync<ResultadoEvaluacion>(cancellationToken: cts.Token)
            ?? throw new InvalidOperationException("Respuesta vacía del Servicio Python");
    }
}
